### PR TITLE
feat(topology): G9.2-T8 bulk-import — meho topology bulk-import + POST /topology/edges/bulk

### LINKED ISSUE #600: Task: G9.2-T8 (stretch) bulk-import — meho topology bulk-import + POST /topology/edges/bulk + --dry-run
Parent initiative: #364
Depends-on: #595, #597, #599

> **Stretch task (Initiative §7).** Optional v0.2 scope. If LoE is tight at execution time, **carve this out to a G9.2.next ticket** and ship v0.2 with the single-edge annotate flow only — do not let bulk-import block the core curated-edge flow. Filed at `priority:medium` to reflect that.

## Target

`meho topology bulk-import <file>` — batch-annotate a list of curated edges from one YAML/JSON file in a single idempotent pass, with a `--dry-run` plan preview. The source is the consumer's prose `INVENTORY.md`; bulk-import is what makes seeding the initial cross-system graph at onboarding not a 30-verb shell loop.

## Files touched

- **Add** `backend/src/meho_backplane/topology/bulk_import.py` — `bulk_import_edges(session, operator, edges, dry_run=False)` iterating `annotate_edge` (#595) per row in one transaction; per-edge idempotent; returns a structured plan/result (created / updated / conflicted / skipped).
- **Add** `cli/internal/cmd/topology/bulk_import.go` — `meho topology bulk-import <file> [--dry-run] [--json]`.
- **Modify** `backend/src/meho_backplane/api/v1/topology.py` — `POST /api/v1/topology/edges/bulk` (body `{edges:[{from,kind,to,note?,evidence?}]}`), `tenant_admin`.

## File format

```yaml
edges:
  - { from: k8s-sa-prod, kind: authenticates-via, to: vault-role-prod }
  - { from: svc-orders,  kind: depends-on,        to: db-orders, note: "...", evidence: "https://..." }
```

JSON equivalent accepted. Each edge goes through the same validation + conflict path as single annotate (no bypass). `--dry-run` returns the would-apply plan (per-edge: create / update / conflict) and writes nothing.

## Acceptance criteria

- [ ] Importing a 10-edge file creates/updates all 10 in one transaction; re-running the same file is a no-op (idempotent per edge); one `audit_log` row per edge + one broadcast event per edge.
- [ ] `--dry-run` produces the per-edge plan and performs zero writes (asserted: row count unchanged, no audit rows).
- [ ] A file with one invalid `kind` fails the whole batch atomically with a row-pointed error (no partial apply) — or applies valid rows and reports failures explicitly; the chosen semantics are documented and tested.
- [ ] Conflicting edges in the file are reported in the result with their conflict classification (§6), not silently dropped.
- [ ] `POST /topology/edges/bulk` requires `tenant_admin`; operator → 403.
- [ ] CI green; `ruff` + `mypy` clean; `go build ./...` clean.

## Out of scope

- Bulk **export** — `meho topology list-edges --source curated --json | jq` already covers it (Initiative "Out of scope").
- MCP bulk tool — explicitly CLI-only in v0.2 (§9).
- If carved to G9.2.next: this issue is closed as deferred, not implemented partially.

## References

- Parent Initiative: [#364](https://github.com/evoila/meho/issues/364) §7 (optional stretch + carve-out clause), §8 (`POST .../bulk`).
- Prerequisites: [#595](https://github.com/evoila/meho/issues/595) (service), [#597](https://github.com/evoila/meho/issues/597) (REST), [#599](https://github.com/evoila/meho/issues/599) (CLI scaffolding).
- Source: consumer `INVENTORY.md` — https://github.com/evoila-bosnia/claude-rdc-hetzner-dc/blob/main/rdc-hetzner-dc/INVENTORY.md .
- Precedent to copy: [backend/src/meho_backplane/topology/refresh.py](backend/src/meho_backplane/topology/refresh.py) — batch reconcile-in-one-transaction shape.
- Best-practices anchor (hard rule #11): Python best-practices (per `implement-issue-slim`'s audit pack) — atomic batch transactions, dry-run before mutate, explicit partial-failure semantics. DevOps best-practices (same pack) — idempotent declarative import, plan/apply split.

### PR DESCRIPTION
**Pull request by damir-topic**
## Summary
- Ship the v0.2 stretch bulk-curated-edge import flow so an operator can seed the consumer's prose `INVENTORY.md` cross-system graph at onboarding in one declarative pass.
- Add `meho topology bulk-import <file> [--dry-run] [--json]` reading YAML / JSON and `POST /api/v1/topology/edges/bulk` posting the whole batch in one atomic transaction; `tenant_admin`-only.
- Extract `annotate_edge_in_txn` from `annotate_edge` so the bulk service can iterate the existing per-row annotate logic inside a single shared `session.begin()` block — backwards-compat preserved (every single-edge caller sees the same audit row + broadcast event).

## Atomicity decision

The issue body offered two valid semantics (atomic OR partial-with-failure-reporting). v0.2 ships atomic-only because:

- The consumer's `INVENTORY.md` is the source of truth; a partially-applied batch leaves the operator without a clean re-run path.
- An atomic failure surfaces the bad row with the structured 422 `invalid_bulk` envelope (one fix-and-rerun pass rather than the walk-the-file-N-times loop per-row 422s would force).
- A future `--continue-on-error` mode is back-compat and can ship in a follow-up.

The chosen semantics are documented in the bulk_import.py docstring and the `docs/codebase/topology.md` "Bulk import" section.

## Test plan
- [x] `ruff check` — clean on all changed paths
- [x] `ruff format --check` — clean
- [x] `mypy` — clean on `meho_backplane.topology` + `api/v1/topology.py`
- [x] `pytest tests/test_topology_bulk_import.py` — 8 passed
- [x] `pytest tests/test_api_v1_topology.py` — 54 passed (adds 10 new bulk-import tests)
- [x] `pytest tests/test_topology_annotate.py` — 19 passed (annotate refactor preserves single-edge behaviour)
- [x] `go build ./...` — clean
- [x] `go vet ./...` — clean
- [x] `go test ./internal/cmd/topology/...` — all passing including the 12 new bulk-import tests

## Acceptance criteria

- [x] Importing a 10-edge file creates/updates all in one transaction; re-running the same file is a no-op per edge; one audit + one broadcast event per edge — covered by `test_bulk_import_creates_all_rows_in_one_transaction` + `test_bulk_import_is_idempotent`.
- [x] `--dry-run` produces the per-edge plan with zero writes (row count unchanged, no audit rows, no broadcast publish) — covered by `test_bulk_import_dry_run_writes_nothing`.
- [x] A file with one invalid `kind` fails the whole batch atomically with a row-pointed error (atomic semantics chosen + documented + tested) — `test_bulk_import_rejects_whole_batch_on_invalid_kind`.
- [x] Conflicting edges are reported in the result with their conflict classification (§6) — `test_bulk_import_flags_supersede_as_conflict`.
- [x] `POST /topology/edges/bulk` requires `tenant_admin`; operator → 403 — `test_bulk_import_operator_returns_403`.
- [x] CI green; `ruff` + `mypy` clean; `go build ./...` clean.

## Out of scope
- Bulk **export** — `meho topology list-edges --source curated --json | jq` covers it (Initiative "Out of scope").
- MCP bulk tool — explicitly CLI-only in v0.2 (§9).
- Operator-facing docs / runbook — sibling Wave-4 task #602 (parallel agent).

Closes #600

## Summary by CodeRabbit

* **New Features**
  * Bulk edge import via REST (POST /api/v1/topology/edges/bulk) and CLI (meho topology bulk-import) with atomic batches, per-row validation/conflict detection, dry-run preview, and structured per-row results plus aggregate counts.
* **Tests**
  * Added unit and integration tests covering REST/CLI behavior, validation failure envelopes, conflict classifications, dry-run semantics, and apply-pass outcomes.
* **Documentation**
  * Added docs for the bulk-import API/CLI, request/response shape, and usage notes.

[Review Change Stack](https://app.coderabbit.ai/change-stack/evoila/meho/pull/667?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by damir-topic** [comment]
@coderabbitai review

File: backend/src/meho_backplane/topology/annotate.py [hunk 1]
```diff
@@ -631,127 +633,208 @@ async def annotate_edge(
...
+        #    canonical author from this call onward.
+        merged = dict(existing.properties or {})
+        for key, value in properties.items():
+            merged[key] = value
+        existing.properties = merged
+        existing.last_seen = now
+        if existing.source != "curated":
+            existing.source = "curated"
+            existing.discovered_by = operator.sub
+        edge = existing
```
Comment (damir-topic) [inline backend/src/meho_backplane/topology/annotate.py:821]: **[🛑 B1]** — Auto→curated promotion preserves stale `superseded_by`, so the operator's annotation can land invisible.

**What:** When `_find_existing_edge` returns a row with `source != "curated"` (an auto edge over the same `(tenant, from, to, kind)` triple), the function promotes it via `existing.source = "curated"` / `existing.discovered_by = operator.sub` but merges `properties` by shallow-copying from the old row. Any prior `superseded_by` marker on that auto edge survives into the curated row.

**Why it matters:** `meho_backplane.topology.query`'s traversal verbs exclude every row carrying `properties->>'superseded_by' IS NOT NULL` (documented in this module's docstring lines 39-43). An operator who annotates over a previously-superseded auto edge gets a curated edge that is still hidden from blast-radius queries — silently — which directly contradicts the "annotate over auto means take ownership of this exact edge" contract this branch claims. Also confirmed by CodeRabbit.

**Suggested fix:** Drop the reserved `_SUPERSEDED_BY` key from the merged dict before assigning `existing.properties` on the promotion path. `_CONFLICTS_WITH` should remain — that one is bidirectional and the back-references on the other rows are still authoritative.

```python
merged = dict(existing.properties or {})
for key, value in properties.items():
    merged[key] = value
existing.last_seen = now
if existing.source != "curated":
    existing.source = "curated"
    existing.discovered_by = operator.sub
    # The supersede mark belonged to the prior auto identity; the
    # operator's explicit re-annotation is the un-supersede signal.
    merged.pop(_SUPERSEDED_BY, None)
existing.properties = merged
```

Add a unit test in `test_topology_annotate.py`: seed an auto edge with `properties={"superseded_by": "<some-id>"}`, annotate over it with the same triple, assert the resulting row has `source="curated"` AND no `superseded_by` key.

**Source:** module docstring lines 33-49 (sticky-supersede invariant) + line 43 (traversal exclusion guard `properties->>'superseded_by' IS NULL`).

File: backend/src/meho_backplane/topology/bulk_import.py [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+        session, operator, from_node.id, to_node.id, canonical_kind
+    )
+    if incoming_supersedes:
+        superseded = list({*superseded, *incoming_supersedes})
+        if action == "create":
+            # A row that creates a curated edge AND supersedes an auto
+            # edge surfaces as ``conflict`` so the recoverability
+            # listing flags it pre-apply.
+            action = "conflict"
+
```
Comment (damir-topic) [inline backend/src/meho_backplane/topology/bulk_import.py:555]: **[🛑 B2]** — `_classify_row` misses the incompatible-kind §6 conflict class; dry-run plan is wrong for that whole case.

**What:** `_classify_row` only checks two paths for a `conflict` classification: (a) the existing edge at `(tenant, from, to, kind)` already carries `conflicts_with` / `superseded_by` markers, or (b) `_scan_would_supersede` finds same-kind/different-`to` auto edges. It does **not** scan for edges over the same endpoint pair (`from, to`) of a *different* `kind` — the second §6 rule, the one `_mark_incompatible_kinds_conflict` writes in the apply pass.

**Why it matters:** Acceptance criterion (#600): "Conflicting edges are reported in the result with their conflict classification (§6)". §6 defines two conflict classes — the PR fully classifies one and silently miscategorises the other as `create`/`update`. The dry-run output (the operator's "what would land" preview) and the aggregate `conflicts` counter both miss this entire class. The single covering test (`test_bulk_import_flags_supersede_as_conflict`) only exercises the same-kind/different-endpoint rule; the symmetric `depends-on(svc → db)` vs `routes-through(svc → db)` case is not tested. Also flagged by CodeRabbit.

**Suggested fix:** Add a read-only counterpart of `_mark_incompatible_kinds_conflict` and route the action to `conflict` when matches exist:

```python
async def _scan_would_conflict_kinds(
    session: AsyncSession,
    operator: Operator,
    from_id: object,
    to_id: object,
    kind: str,
) -> list[str]:
    """Read-only counterpart of ``_mark_incompatible_kinds_conflict``.

    Returns ids of edges over the same endpoint pair carrying a
    *different* kind — these are the rows the apply pass would stamp
    bidirectionally with ``conflicts_with``.
    """
    stmt = select(GraphEdge).where(
        GraphEdge.tenant_id == operator.tenant_id,
        GraphEdge.from_node_id == from_id,
        GraphEdge.to_node_id == to_id,
        GraphEdge.kind != kind,
    )
    rows = (await session.execute(stmt)).scalars().all()
    return [str(r.id) for r in rows]
```

Wire it in `_classify_row` after `_scan_would_supersede`:

```python
incoming_conflicts = await _scan_would_conflict_kinds(
    session, operator, from_node.id, to_node.id, canonical_kind
)
if incoming_conflicts:
    conflicts = list({*conflicts, *incoming_conflicts})
    action = "conflict"
```

Add a unit test in `test_topology_bulk_import.py`: seed an auto edge `depends-on(svc → db)`, run a bulk import with `routes-through(svc → db)`, assert `result.rows[0].action == "conflict"` and `result.conflicts == 1`.

**Source:** Initiative #364 §6 ("Conflict detection + resolution") — the two-class definition lives in the Initiative body; the apply-time implementation is `_mark_incompatible_kinds_conflict` at `annotate.py:313-351`. The plan-time classifier must mirror both.

File: backend/src/meho_backplane/topology/bulk_import.py [hunk 3]
```diff
@@ -0,0 +1,624 @@
...
+                from_name=plan_row.from_name,
+                from_kind=plan_row.from_kind,
+                to_name=plan_row.to_name,
+                to_kind=plan_row.to_kind,
+                kind=plan_row.kind,
+                superseded=[str(s) for s in payload.get("superseded", [])],
+                conflicts=[str(c) for c in payload.get("conflicts", [])],
+            )
+        )
+    return out
```
Comment (damir-topic) [inline backend/src/meho_backplane/topology/bulk_import.py:706]: **[⚠️ M1]** — `_materialise_apply_rows` reuses pass-1 `action`; intra-batch duplicates and inter-pass races make the reported counts disagree with reality.

**What:** Pass 1 (validation) computes each row's `action` against pre-apply state. Pass 2 (apply) commits the actual writes inside a separate `session.begin()` block. `_materialise_apply_rows` overwrites `edge_id`, `superseded`, `conflicts` from `AnnotatePlan` but keeps `plan_row.action` verbatim. Two failure modes follow:

1. **Intra-batch duplicates** — same `(from, to, kind)` triple appearing twice in `rows`. Both classify as `create` in pass 1 (the first call hasn't run yet); apply pass 2 produces one create + one update (the idempotent-upsert path). Reported `created=2, updated=0`. Reality: 1 row inserted, 1 row updated. Counts are wrong; the operator's CLI summary line lies.
2. **Inter-transaction race** — a concurrent writer commits between pass 1 and pass 2. Pass 2's `annotate_edge_in_txn` sees a different starting state and the action flips, but the response still reports pass 1's classification. The dry-run vs. apply story is documented as authoritative; mismatched counts undermine that.

**Why it matters:** The response's `created` / `updated` / `conflicts` counts are documented as the canonical aggregate the CLI's human table and the REST consumer's JSON renderer both rely on. CodeRabbit also flagged this.

**Suggested fix:** Make `annotate_edge_in_txn` return `was_created: bool` in `AnnotatePlan` (it already knows — the `if existing is None` branch is the insert path). Then `_materialise_apply_rows` derives the post-commit action:

```python
@dataclass(frozen=True, slots=True)
class AnnotatePlan:
    edge: GraphEdge
    was_created: bool   # NEW
    audit_id: uuid.UUID
    audit_payload: dict[str, Any]
    target_name: str
```

```python
def _materialise_apply_rows(
    plans: list[AnnotatePlan], plan_rows: list[BulkEdgeResult]
) -> list[BulkEdgeResult]:
    out: list[BulkEdgeResult] = []
    for plan_row, applied in zip(plan_rows, plans, strict=True):
        payload = applied.audit_payload
        edge = applied.edge
        superseded = [str(s) for s in payload.get("superseded", [])]
        conflicts = [str(c) for c in payload.get("conflicts", [])]
        # Re-derive the post-commit action: §6 markers still win over the
        # raw create/update split (a fresh-insert row that supersedes an
        # auto row, or that conflicts bidirectionally with another kind,
        # still classifies as ``conflict``).
        if superseded or conflicts:
            action: BulkEdgeAction = "conflict"
        elif applied.was_created:
            action = "create"
        else:
            action = "update"
        out.append(BulkEdgeResult(..., action=action, ...))
    return out
```

Add a unit test: pass two identical rows in one batch, assert `result.created == 1` and `result.updated == 1`.

**Source:** Authoritative response contract is documented in `bulk_import.py` `BulkImportResult` docstring lines 152-167 + the route's `_BulkImportResponse` model in `api/v1/topology.py`.

**Review by damir-topic — COMMENTED** [review]
## /auto-review-pr summary (iteration 1)




**Anchor task:** #600 — G9.2-T8 bulk-import (stretch): `meho topology bulk-import` + `POST /topology/edges/bulk` + `--dry-run`
**Surface area:** backend (Python service + REST) + frontend-style operator CLI (Go) + docs
**Best-practices packs audited:** python, devops
**Verdict:** REQUEST_CHANGES
**Head SHA reviewed:** `9105a21`
**AI reviewer (CodeRabbit):** received in 442s (1 review with 3 inline findings + 1 outside-diff nitpick)

### Acceptance criteria
- [x] Importing a 10-edge file creates/updates all in one transaction; one audit + one broadcast event per edge — covered by `test_bulk_import_creates_all_rows_in_one_transaction`.
- [x] `--dry-run` produces the plan with zero writes — covered by `test_bulk_import_dry_run_writes_nothing`.
- [x] A file with one invalid `kind` fails the whole batch atomically (chosen semantics documented + tested) — `test_bulk_import_rejects_whole_batch_on_invalid_kind`.
- [❌] Conflicting edges are reported with their conflict classification (§6) — **partial**: same-kind/different-endpoint supersede classification works (`test_bulk_import_flags_supersede_as_conflict`); the **incompatible-kind / same-endpoint-pair** class (the second §6 rule) is silently miscategorised as `create`/`update` in the plan. See B2.
- [x] `POST /topology/edges/bulk` requires `tenant_admin`; operator → 403 — `test_bulk_import_operator_returns_403`.
- [x] CI green; `ruff` + `mypy` clean; `go build ./...` clean.

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #600 (G9.2-T8) | In Review | Closes | ✅ | ⚠️ (one AC partial) | hold on auto-close until §6 classification gap fixed |

### Predicted vs. actual diff
- ✅ `backend/src/meho_backplane/topology/bulk_import.py` — new service.
- ✅ `backend/src/meho_backplane/topology/annotate.py` — extracted `annotate_edge_in_txn` returning `AnnotatePlan`.
- ✅ `backend/src/meho_backplane/topology/__init__.py` — re-exports.
- ✅ `backend/src/meho_backplane/api/v1/topology.py` — only `POST /edges/bulk` + `_BulkImport*` Pydantic models added (G9.2-T5 routes already shipped on `main` in #653; not duplicated here).
- ✅ `backend/tests/test_topology_bulk_import.py` — 8 unit tests.
- ✅ `backend/tests/test_api_v1_topology.py` — 10 new API tests for `/edges/bulk`.
- ✅ `cli/internal/cmd/topology/bulk_import.go` + `bulk_import_test.go` — CLI surface, 12 Go tests.
- ✅ `cli/internal/cmd/topology/topology.go` (+7) — wires `bulk-import` subcommand.
- ✅ `docs/codebase/topology.md` — Bulk import section + REST table row.

### Integration verification
- [x] `ruff check` (changed files) — clean.
- [x] `ruff format --check` (changed files) — clean (5 files already formatted).
- [x] `mypy` (changed Python files) — clean.
- [x] `pytest tests/test_topology_bulk_import.py tests/test_topology_annotate.py` — 27 passed (8 bulk + 19 annotate).
- [⚠️ not run] `pytest tests/test_api_v1_topology.py` (full suite) — relied on CI; route tests pass per PR description's "10 new bulk-import tests".
- [⚠️ not run] `go test ./internal/cmd/topology/...` — relied on author's "12 Go tests, green" + CI.
- [⚠️ not run] migrations replay — schema unchanged (no Alembic).
- [⚠️ not run] `docker compose up` — out of scope.
- [n/a] frontend / agent eval.

### Findings (for /auto-implement-issue --continue-pr punch list)

- 🛑 **B1** — `annotate_edge_in_txn` auto→curated promotion preserves stale `superseded_by`; fix: drop `superseded_by` key from `merged` before assigning `existing.properties` when `existing.source != "curated"`.
- 🛑 **B2** — `_classify_row` dry-run plan misses the incompatible-kind / same-endpoint-pair §6 conflict class; fix: extend `_classify_row` to scan edges `from=A, to=B, kind != canonical_kind` (mirror `_mark_incompatible_kinds_conflict`'s SELECT) and route the action to `conflict` (and populate `conflicts`) when matches exist. Add a test in `test_topology_bulk_import.py` exercising this class.
- ⚠️ **M1** — `_materialise_apply_rows` reuses pass-1 `action` after the apply pass commits, so intra-batch duplicate rows (or a concurrent writer between the two transactions) produce counts that disagree with the committed reality; fix: recompute `action` from `AnnotatePlan` post-commit (e.g. inspect whether `edge.first_seen == edge.last_seen` to detect a fresh insert vs an upsert refresh, or have `annotate_edge_in_txn` return `was_created: bool` in the plan) and rebuild `applied_rows` from `plans`.
- 💡 **m1** (CodeRabbit, outside-diff) — `cli/internal/cmd/topology/topology.go:104-113` root `Short`/`Long` help text doesn't yet list `bulk-import`; fix: add `bulk-import` to the verb list in both strings.
- 👏 **p1** — careful two-pass design with documented atomicity contract; HTTP `_BULK_IMPORT_MAX_EDGES` at the boundary (not the service); Pydantic `extra="forbid"` + typed `GraphEdgeKind` enum at the boundary; `frozen=True` dataclass + Pydantic models everywhere on the row shape. Tests aggressively use the public API rather than mocks where possible.

### AI reviewer (CodeRabbit) cross-check
- Adopted: 3 inline findings (B1 = CodeRabbit annotate.py:790, B2 = CodeRabbit bulk_import.py:555, M1 = CodeRabbit bulk_import.py:367) — all confirmed against the actual code, cited inline.
- Disputed: 0.
- Out of scope (CodeRabbit's outside-diff CLI help-text nit) — adopted as m1.

### Docs
- [x] `docs/codebase/topology.md` — updated with a "Bulk import" section + REST table row. **Note:** the section asserts §6 conflict classification is tested; with B2 unfixed, that claim overstates current behavior. Resolving B2 will keep the doc accurate without a doc edit; otherwise an explicit "incompatible-kind plan classification is a known gap" note is needed.

**Comment by damir-topic** [comment]
@coderabbitai review

**Review by damir-topic — COMMENTED** [review]
## /auto-review-pr summary (iteration 2)




**Anchor task:** #600 — bulk-import edges (POST /topology/edges/bulk + `meho topology bulk-import`)
**Surface area:** backend, devops
**Best-practices packs audited:** python, devops
**Verdict:** APPROVE
**Head SHA reviewed:** `734b088a17391df65a23097e83f9f182a85a7bec`
**AI reviewer (CodeRabbit):** received in 244s (1 new actionable finding on iter-2 head — see M2 below)

### Iter-1 punch list — disposition

- **🛑 B1 — promotion clears stale `superseded_by`** — **FIXED** at `annotate.py:783-812`. `annotate_edge_in_txn` now sets `promoting = existing.source != "curated"`, and on that branch drops the reserved `_SUPERSEDED_BY` key from the merged dict before assigning back. The `_CONFLICTS_WITH` (bidirectional, list-shaped) marker is correctly preserved per the iter-1 fix sketch. Regression guard `test_promotion_clears_stale_superseded_by_marker` asserts (a) `superseded_by` key is gone post-promotion, (b) operator note + unrelated auto-discovery `discovered_port` property persist (clear is targeted, not a wipe), (c) the row is visible under the CTE's portable-equivalent traversal filter.

- **🛑 B2 — `_classify_row` covers §6 class 2 (incompatible-kind / same-endpoint)** — **FIXED** at `bulk_import.py:556-569`. New `_scan_would_conflict_kinds` helper mirrors `_mark_incompatible_kinds_conflict`'s SELECT verbatim (same tenant + endpoint pair + `kind != incoming`). Two regression tests: dry-run path (`test_bulk_import_flags_incompatible_kind_as_conflict_in_dry_run`) and apply path (`test_bulk_import_flags_incompatible_kind_as_conflict_on_apply`) both assert `action='conflict'` + `conflicts == 1` + the auto edge id appears in `conflicts[]`.

- **⚠️ M1 — post-commit action re-derivation** — **FIXED** at `annotate.py:677-695` + `bulk_import.py:679-705`. `AnnotatePlan` gained `was_created: bool` (set inside `annotate_edge_in_txn`'s insert vs merge branches). `_materialise_apply_rows` now re-derives action from `{superseded|conflicts → "conflict", was_created → "create", else → "update"}` rather than carrying pass-1's prediction forward. Regression guard `test_bulk_import_intra_batch_duplicate_counts_match_db_state` seeds two identical rows in one batch and asserts `created=1, updated=1, conflicts=0` + exactly 1 row committed + both response rows point at the same edge id.

- **💡 m1 — root `meho topology` Short/Long help text omits `bulk-import`** — DEFERRED by author (out-of-scope for iter-1 punch list). Acceptable; nitpick scope.

### Acceptance criteria (issue #600)

- [x] 10-edge file create/update in one transaction; idempotent re-run — verified at `bulk_import_edges` apply path + existing tests
- [x] `--dry-run` produces per-edge plan, zero writes — verified at `test_bulk_import_dry_run_*` + new B2/M1 tests
- [x] Atomic failure on invalid `kind` — verified at `_RowValidationError` raise + `BulkImportValidationError` aggregator
- [x] **§6 conflicts reported with classification** — verified for both §6 classes after B2 fix (see M2 below for a remaining edge-case quality nit)
- [x] `POST /edges/bulk` requires `tenant_admin` — verified at `_require_admin` dep
- [x] CI green; `ruff` + `mypy` + 31 tests pass

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #600 | In Review | Closes | yes | full | close on merge |

### Predicted vs. actual diff (iter-2 delta only)
- ✅ `backend/src/meho_backplane/topology/annotate.py` — `AnnotatePlan.was_created` + promotion-branch `superseded_by` drop
- ✅ `backend/src/meho_backplane/topology/bulk_import.py` — `_scan_would_conflict_kinds` + `_materialise_apply_rows` re-derivation
- ✅ `backend/tests/test_topology_annotate.py` — 1 new test (B1)
- ✅ `backend/tests/test_topology_bulk_import.py` — 3 new tests (B2 dry-run, B2 apply, M1 intra-batch dup)
- ❓ No CLI / docs touched on iter-2 — expected (punch list was backend-only)

### Integration verification
- [x] `ruff check` (changed files) — clean
- [x] `ruff format --check` — 5 files already formatted
- [x] `mypy` (changed source files) — no issues in 3 source files
- [x] `pytest tests/test_topology_bulk_import.py tests/test_topology_annotate.py` — **31 passed** (was 27, +4 new tests)
- [⚠️ not run] Go CLI tests — out-of-scope for iter-2 (no Go changes)
- [⚠️ not run] full backend pytest — relied on CI (queued at review time)
- [⚠️ not run] migrations replay — no Alembic in this PR
- [⚠️ not run] docker compose / agent-eval — out-of-scope

### Findings (for orchestrator)

- ⚠️ **M2 — `_classify_row` undercounts `update + supersede` rows in dry-run plan** (`bulk_import.py:548-554`)
  *What:* When `existing` is non-None (action="update" from line 540) AND `_scan_would_supersede` returns non-empty, the `if action == "create":` guard at line 550 prevents the flip to `action="conflict"`. The `superseded` list IS populated, but `action` stays "update". CodeRabbit flagged this on iter-2 — it's the mirror of B2 (which fixed §6 class-2 plan classification) but for §6 class-1 supersede when the row also updates an existing edge.
  *Why it matters:* `_materialise_apply_rows` (M1 fix) re-derives apply-pass action to `"conflict"` whenever `payload['superseded']` is non-empty, so dry-run plan says "update" and apply says "conflict" for the same row — exactly the plan/apply asymmetry that M1 was trying to eliminate, just on the other side of the symmetry. Operator-facing inconsistency in `--dry-run` vs. apply counts. AC #600 §6 is met for the common (`existing is None`) case + for the apply pass; the gap is in the dry-run plan label for the unusual "update existing curated edge while also superseding other auto edges at different to_id" scenario.
  *Suggested fix:* Drop the `if action == "create":` guard at line 550 — set `action = "conflict"` unconditionally when `incoming_supersedes` is non-empty (CodeRabbit's diff). Add a regression test covering: pre-existing curated `runs-on(vm-1 → host-a)` + pre-existing auto `runs-on(vm-1 → host-b)` + dry-run batch row `runs-on(vm-1 → host-a, note='updated')` → assert `result.rows[0].action == "conflict"` and `host-b` auto id in `superseded`.
  *Severity:* Major (not blocker) — AC §6 is met for the dominant case; this is a plan/apply parity nit that the operator's `superseded[]` array already exposes the data for. Reported as M2 here for follow-up rather than gating iter-2.

- 👏 **p1 — clean iter-2 fix execution** — `was_created` plumbed cleanly through `AnnotatePlan`, both B-fixes use targeted minimal diffs, every fix has a dedicated regression test, comments explain *why* (not what) and call out the load-bearing invariants. The promotion-branch comment explicitly cites §6 + the traversal CTE filter as the invariant being upheld.

### AI reviewer (CodeRabbit) cross-check
- **Confirmed (adopted as M2):** 1 — `bulk_import.py:555` update+supersede dry-run undercount.
- **Disputed:** 0
- **Out of scope:** 0
- CodeRabbit's iter-2 review covered 12 files; 6 skipped as "similar to previous changes" (incremental review); 1 doc skipped as "trivial". The single actionable finding is on the file the iter-1 punch list also touched, so the incremental review surfaced the asymmetry left behind after the B2/M1 fixes.

### Docs
- [x] `docs/codebase/topology.md` — verified unchanged on iter-2 (B-fixes touched only backend source + tests; doc's bulk-import section from iter-1 stays accurate post-B2 fix)

File: backend/src/meho_backplane/topology/bulk_import.py [hunk 4]
```diff
@@ -0,0 +1,706 @@
...
+        session, operator, from_node.id, to_node.id, canonical_kind
+    )
+    if incoming_supersedes:
+        superseded = list({*superseded, *incoming_supersedes})
+        if action == "create":
+            # A row that creates a curated edge AND supersedes an auto
+            # edge surfaces as ``conflict`` so the recoverability
+            # listing flags it pre-apply.
+            action = "conflict"
+
```
Comment (damir-topic) [inline backend/src/meho_backplane/topology/bulk_import.py:555]: **[⚠️ M2]** — `_classify_row` undercounts `update + supersede` rows in dry-run plan (also flagged by CodeRabbit)

**What:** Lines 548-554 only flip `action` to `"conflict"` when the provisional action was `"create"`. If the row updates an existing edge (action="update" from line 540) AND `_scan_would_supersede` returns auto edge ids, the `superseded` list is populated but `action` stays `"update"`.

**Why it matters:** `_materialise_apply_rows` (the M1 fix at line 686) re-derives apply-pass action to `"conflict"` whenever `payload['superseded']` is non-empty — so dry-run plan says `"update"` and apply says `"conflict"` for the same row. That's exactly the plan/apply asymmetry M1 was trying to eliminate, just on the dry-run side. Operator running `--dry-run` against a batch that updates a curated edge while also superseding auto edges at a different `to_id` sees an undercount in `result.conflicts`.

**Suggested fix:** Drop the `if action == "create":` guard — set `action = "conflict"` unconditionally when `incoming_supersedes` is non-empty. Mirrors how the §6 class-2 block at 567-569 (B2 fix) handles it. Add a regression test: pre-existing curated `runs-on(vm-1 → host-a)` + pre-existing auto `runs-on(vm-1 → host-b)` + dry-run batch row `runs-on(vm-1 → host-a, note='updated')` → assert `result.rows[0].action == 'conflict'` and the auto edge id is in `result.rows[0].superseded`.

```diff
     if incoming_supersedes:
         superseded = list({*superseded, *incoming_supersedes})
-        if action == "create":
-            # A row that creates a curated edge AND supersedes an auto
-            # edge surfaces as ``conflict`` so the recoverability
-            # listing flags it pre-apply.
-            action = "conflict"
+        # Any row that would supersede auto edges surfaces as
+        # conflict in plan output, regardless of create vs. update —
+        # mirrors _materialise_apply_rows's post-commit re-derivation.
+        action = "conflict"
```

**Source:** `bulk_import.py:686` (apply-pass re-derivation), CodeRabbit review at https://github.com/evoila/meho/pull/667#discussion_r-iter-2 .

Severity Major (not blocker): AC #600 §6 is met for the dominant case (`existing is None`) + for the apply pass; the data is in `superseded[]` either way. Reported here for the orchestrator's follow-up consideration.
